crates/host/src/cache.rs

```diff
@@ -0,0 +1,588 @@
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    internal: Arc<RwLock<CacheInternal<K, V>>>,
+    purge_cycles: Option<u32>,
+    purge_cycle_count: AtomicU32,
+}
+
+impl<K, V> Cache<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    /// Create a new cache with the specified TTL
+    pub fn new(ttl: Duration) -> Self {
+        Self {
+            internal: Arc::new(RwLock::new(CacheInternal::new(ttl))),
+            purge_cycles: None,
+            purge_cycle_count: AtomicU32::new(0),
+        }
+    }
+
+    /// Create a new cache builder for configuring cache settings
+    pub fn builder() -> CacheBuilder<K, V> {
+        CacheBuilder::new()
+    }
+
+    /// Insert a key-value pair into the cache
+    pub async fn insert(&self, key: K, value: V) {
+        self.check_purge().await;
+        self.internal.write().await.insert(key, value);
+    }
+
+    /// Get a value from the cache, returning None if expired or not found
+    pub async fn get(&self, key: &K) -> Option<V> {
+        self.internal.write().await.get(key)
+    }
+
+    /// Remove a key-value pair from the cache
+    pub async fn remove(&self, key: &K) {
+        self.internal.write().await.remove(key);
+    }
+
+    /// Manually trigger a single purge cycle to remove expired entries
+    pub async fn poll_purge_once(&self) {
+        let mut cache = self.internal.write().await;
+        std::future::poll_fn(|cx| cache.poll_purge(cx)).await;
+        self.purge_cycle_count.store(0, Ordering::Relaxed);
+    }
+
+    async fn check_purge(&self) {
+        if let Some(purge_cycles) = self.purge_cycles {
+            self.purge_cycle_count.fetch_add(1, Ordering::Relaxed);
+            if self.purge_cycle_count.load(Ordering::Relaxed) >= purge_cycles {
+                self.poll_purge_once().await;
+            }
+        }
+    }
+}
+
+impl<K, V> std::fmt::Debug for Cache<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Cache")
+            .field("purge_cycles", &self.purge_cycles)
+            .field(
+                "purge_cycle_count",
+                &self.purge_cycle_count.load(Ordering::Relaxed),
+            )
+            .finish()
+    }
+}
+
+impl<K, V> Clone for Cache<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn clone(&self) -> Self {
+        Self {
+            internal: Arc::clone(&self.internal),
+            purge_cycles: self.purge_cycles,
+            purge_cycle_count: AtomicU32::new(self.purge_cycle_count.load(Ordering::Relaxed)),
```

Comment:
The `purge_cycle_count` should probably be in an `Arc` so it is cloned and shared as well. Otherwise, clones of the cache will be purging on different cycles

Reply:
You mean if it's modified? Later on? But sure, I can update it, being a u32, it copies right? But you are right, if anyone modifies it, it will stay consistent fixing....!!!